### PR TITLE
Bugfix: Implement roll control on Solo gimbal

### PR DESCRIPTION
**Pull request by igorsgcampos**
I was testing https://github.com/ArduPilot/ardupilot/pull/15878 and noticed that Solo was not changing the roll angle of the gimbal according to what was commanded. Upon inspecting the code, I saw that it just ignored the roll input, despite the vehicle's capability of processing the command.

This fixes the lack of roll control for Solo, making it compliant with the mavlink DO_MOUNT_CONTROL specification.

### CODE REVIEW

**Comment by igorsgcampos** [comment]
I asked to put this on the dev-call, but it was forgotten, I would appreciate if it could get reviewed, it is pretty straight-forward.

**Review by khancyr — APPROVED** [review]
I don't have SOLO, so I cannot check. 
The code is mostly a duplicate from the tilt code except on L399 where the angle limit are adapted to roll.

**Comment by khancyr** [comment]
@igorsgcampos thanks for this, it looks good to me. Could you fix the commit message ?
it should be `AP_Mount: Implement roll control on Solo gimbal` 
I have tag this for the next devcall so it should be merged fast.

**Comment by igorsgcampos** [comment]
Done, it is updated now!

**Comment by khancyr** [comment]
Thanks, hopefully it should be merged on next devcall.

**Comment by rmackay9** [comment]
@igorsgcampos,

You've tested this change on a Solo gimbal?  The solo gimbal has very limited roll movement I think so we really should test it on real hardware to make sure that the change is useful.

**Comment by igorsgcampos** [comment]
Yes, I've tested it. The specs claim 40 degrees, that is why I limited to 30.

**Comment by tridge** [comment]
I've asked @Pedals2Paddles and @davidbuzz to check this

**Comment by rmackay9** [comment]
This looks good to me.

**Comment by davidbuzz** [comment]
I have no particular opinion on this, but if its been tested on a real solo gimbal, it sounds ok it theory.   lets see what @Pedals2Paddles  says.

**Comment by Pedals2Paddles** [comment]
I should be able to test this week. What is the functional purpose of this change? It says bug fix but I'm not aware of very being a bug requiring this? In other words what problem are we solving here?

**Comment by peterbarker** [comment]
On Mon, 11 Jan 2021, Matt wrote:

> I should be able to test this week. What is the functional purpose of this change? It says bug fix but I'm not aware of very being a bug requiring this?
> In other words what problem are we solving here?


Not a bug fix.

With the apprioriate changes made you should be able to deliverately roll 
the camera.

**Comment by igorsgcampos** [comment]
> I should be able to test this week. What is the functional purpose of this change? It says bug fix but I'm not aware of very being a bug requiring this? In other words what problem are we solving here?

The purpose is enabling #15878 to work.

It is a bugfix in the sense that presently, if the Solo is issued a DO_MOUNT_CONTROL, it will ignore the roll angle, thus presenting a different behavior than the protocol specification. Sorry if the description was confusing according to your standards, but I considered this deviation from the protocol a bug.

**Comment by igorsgcampos** [comment]
Here is a log from a flight using this feature:

[log_84_2020-12-20-16-08-56.zip](https://github.com/ArduPilot/ardupilot/files/5802437/log_84_2020-12-20-16-08-56.zip)

**Comment by igorsgcampos** [comment]
This overview of part of the images it took on the flight should demonstrate the behavior of the camera changing its roll:

image

**Comment by Pedals2Paddles** [comment]
Ah ok. That's pretty cool.  I believe you're probably the first person who's ever actually *wanted* Solo gimbal roll control outside of automated level stabilization.  Or at least the first to say something anyway.  I would consider this an enhancement rather than a bugfix actually.

I'm not equipped to test surveys like you are, nor do I have experience with it.  So I will defer to your expertise and abilities on it doing what you're trying to achieve there. Is the roll angle something your sending in a command from the survey mission?  When the mission is done or you're otherwise just wanting it to be level, does it go back normal operation remaining level?

**Comment by igorsgcampos** [comment]
>Is the roll angle something your sending in a command from the survey mission?

Right now I implemented a new command, which will eventually move from AP_Camera to AP_Vehicle. The implementation picks the AP_Mount singleton and instructs the camera to roll left and right according to the command parameters. For reference:
https://github.com/mavlink/mavlink/pull/1508

>When the mission is done or you're otherwise just wanting it to be level, does it go back normal operation remaining level?

At the end, it remains at the last position, but I changed QGC so that it adds a trailing DO_MOUNT_CONTROL with orientations 0,0,0, thus making the camera point forward.

**Comment by Pedals2Paddles** [comment]
Understood. IIRC, nothing else is sending it roll commands so the general users with a controller flying around with just the pitch control will not see any difference?  If so, then I am happy with your testing/implementation and will approve the PR review.

**Comment by igorsgcampos** [comment]
> Understood. IIRC, nothing else is sending it roll commands so the general users with a controller flying around with just the pitch control will not see any difference? If so, then I am happy with your testing/implementation and will approve the PR review.

That is correct, it should not affect present functions, only enable roll control, which is currently unused.

**Review by Pedals2Paddles — APPROVED** [review]
Fantastic.  Nice work and thank you for the contribution back to the community with this.  I approve. @davidbuzz can you approve your review too if you do?

**Comment by peterbarker** [comment]
Merged, thanks @igorsgcampos 

On the "bugfix" thing - we deviate from the MAVLink standard a *lot*, because the standard is silly in many places.... so that deviation isn't necessarily the bug :-)

**Review by davidbuzz — COMMENTED** [review]
Ok

**Comment by igorsgcampos** [comment]
> On the "bugfix" thing - we deviate from the MAVLink standard a _lot_, because the standard is silly in many places.... so that deviation isn't necessarily the bug :-)

I didn't mean the MAVLink protocol in this case, but rather ArduPilot's documentation: https://ardupilot.org/copter/docs/mission-command-list.html#do-mount-control

In this specific case, Solo was behaving different from other Copters, but I do understand that this might be a corner case and unlikely to affect the vast majority of users.

Thanks for merging this!

**Pull request merged**
